File: lovart_reverse/mcp/configure.py
"""Configure local MCP clients to use the Lovart MCP server."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from lovart_reverse.errors import LovartError


SUPPORTED_MCP_CLIENTS = ("codex", "claude", "opencode", "openclaw")
MANAGED_MARKER = "# Managed by lovart-reverse"


class ConfigConflictError(LovartError):
    def __init__(self, message: str, details: dict[str, Any] | None = None):
        super().__init__("config_conflict", message, details or {}, 2)


@dataclass(frozen=True)
class McpClientContext:
    lovart_path: Path
    home: Path
    dry_run: bool = False
    yes: bool = False
    force: bool = False


def mcp_status(*, clients: str = "auto", lovart_path: Path | None = None, home: Path | None = None) -> dict[str, Any]:
    ctx = McpClientContext(_lovart_path(lovart_path), _home(home), dry_run=True)
    selected = _select_mcp_clients(clients, ctx, include_missing=True)
    statuses = [_status_for(client, ctx) for client in selected]
    return {"lovart_path": str(ctx.lovart_path), "clients": statuses}


def mcp_install(
    *,
    clients: str = "auto",
    lovart_path: Path | None = None,
    home: Path | None = None,
    dry_run: bool = False,
    yes: bool = False,
    force: bool = False,
) -> dict[str, Any]:
    ctx = McpClientContext(_lovart_path(lovart_path), _home(home), dry_run=dry_run, yes=yes, force=force)
    selected = _select_mcp_clients(clients, ctx, include_missing=False)
    if clients == "none":
        selected = []
    results = [_install_client(client, ctx) for client in selected]
    return {
        "lovart_path": str(ctx.lovart_path),
        "dry_run": dry_run,
        "force": force,
        "mcp_clients_requested": clients,
        "mcp_clients_selected": selected,
        "results": results,
    }


def _lovart_path(path: Path | None) -> Path:
    if path:
        return path.expanduser().resolve()
    if getattr(sys, "frozen", False):
        return Path(sys.executable).resolve()
    argv0 = Path(sys.argv[0])
    if argv0.exists():
        return argv0.expanduser().resolve()
    found = shutil.which("lovart")
    return Path(found).resolve() if found else argv0


def _home(path: Path | None) -> Path:
    return (path or Path.home()).expanduser().resolve()


def _select_mcp_clients(clients: str, ctx: McpClientContext, *, include_missing: bool) -> list[str]:
    if clients == "none":
        return []
    if clients == "all":
        return list(SUPPORTED_MCP_CLIENTS)
    if clients == "auto":
        selected = [client for client in SUPPORTED_MCP_CLIENTS if _client_detected(client, ctx)]
        if selected:
            return selected
        return list(SUPPORTED_MCP_CLIENTS) if include_missing else ["codex"]
    selected = [item.strip().lower() for item in clients.split(",") if item.strip()]
    unknown = sorted(set(selected) - set(SUPPORTED_MCP_CLIENTS))
    if unknown:
        raise LovartError("input_error", "unknown MCP client", {"clients": unknown, "supported_clients": list(SUPPORTED_MCP_CLIENTS)}, 2)
    return selected


def _client_detected(client: str, ctx: McpClientContext) -> bool:
    if client == "codex":
        return (ctx.home / ".codex").exists()
    if client == "claude":
        return shutil.which("claude") is not None
    if client == "opencode":
        return shutil.which("opencode") is not None or (ctx.home / ".config" / "opencode").exists()
    if client == "openclaw":
        return shutil.which("openclaw") is not None
    return False


def _status_for(client: str, ctx: McpClientContext) -> dict[str, Any]:
    if client == "codex":
        return _codex_status(ctx)
    if client == "claude":
        return _command_status("claude", ["claude", "mcp", "add", "--transport", "stdio", "lovart", "--scope", "user", "--", str(ctx.lovart_path), "mcp"])
    if client == "opencode":
        return _opencode_status(ctx)
    if client == "openclaw":
        return _command_status("openclaw", ["openclaw", "mcp", "set", "lovart", _openclaw_payload(ctx)])
    raise LovartError("input_error", "unknown MCP client", {"client": client}, 2)


def _install_client(client: str, ctx: McpClientContext) -> dict[str, Any]:
    if client == "codex":
        return _install_codex(ctx)
    if client == "claude":
        return _install_cli_client("claude", ["claude", "mcp", "add", "--transport", "stdio", "lovart", "--scope", "user", "--", str(ctx.lovart_path), "mcp"], ctx)
    if client == "opencode":
        return _install_opencode(ctx)
    if client == "openclaw":
        return _install_cli_client("openclaw", ["openclaw", "mcp", "set", "lovart", _openclaw_payload(ctx)], ctx)
    raise LovartError("input_error", "unknown MCP client", {"client": client}, 2)


def _codex_config_path(ctx: McpClientContext) -> Path:
    return ctx.home / ".codex" / "config.toml"


def _codex_block(ctx: McpClientContext) -> str:
    return "\n".join(
        [
            MANAGED_MARKER,
            "[mcp_servers.lovart]",
            f'command = "{_toml_string(str(ctx.lovart_path))}"',
            'args = ["mcp"]',
            "",
        ]
    )


def _codex_status(ctx: McpClientContext) -> dict[str, Any]:
    path = _codex_config_path(ctx)
    text = _read_text(path)
    configured = "[mcp_servers.lovart]" in text and str(ctx.lovart_path) in text and 'args = ["mcp"]' in text
    managed = MANAGED_MARKER in text and "[mcp_servers.lovart]" in text
    return {"client": "codex", "type": "file", "path": str(path), "exists": path.exists(), "configured": configured, "managed": managed}


def _install_codex(ctx: McpClientContext) -> dict[str, Any]:
    path = _codex_config_path(ctx)
    text = _read_text(path)
    has_block = "[mcp_servers.lovart]" in text
    managed = MANAGED_MARKER in text and has_block
    if has_block and not managed and not ctx.force:
        raise ConfigConflictError("existing unmanaged Codex Lovart MCP config", {"client": "codex", "path": str(path), "recommended_actions": ["rerun with --force", "edit the config manually"]})
    next_text = _replace_toml_lovart_block(text, _codex_block(ctx))
    return _write_config_result("codex", path, next_text, ctx, preview={"toml": _codex_block(ctx)})


def _replace_toml_lovart_block(text: str, block: str) -> str:
    lines = text.splitlines()
    start = next((index for index, line in enumerate(lines) if line.strip() == "[mcp_servers.lovart]"), None)
    if start is None:
        prefix = text.rstrip()
        return (prefix + "\n\n" if prefix else "") + block
    if start > 0 and lines[start - 1].strip() == MANAGED_MARKER:
        start -= 1
    end = start + 1
    while end < len(lines):
        stripped = lines[end].strip()
        if stripped.startswith("[") and stripped.endswith("]") and stripped != "[mcp_servers.lovart]":
            break
        end += 1
    new_lines = lines[:start] + block.rstrip().splitlines() + lines[end:]
    return "\n".join(new_lines).rstrip() + "\n"


def _opencode_config_path(ctx: McpClientContext) -> Path:
    return ctx.home / ".config" / "opencode" / "opencode.json"


def _opencode_status(ctx: McpClientContext) -> dict[str, Any]:
    path = _opencode_config_path(ctx)
    data = _read_json(path)
    config = (data.get("mcp") or {}).get("lovart") if isinstance(data, dict) else None
    configured = isinstance(config, dict) and config.get("command") == [str(ctx.lovart_path), "mcp"] and config.get("enabled") is True
    managed = isinstance(config, dict) and config.get("managed_by") == "lovart-reverse"
    return {"client": "opencode", "type": "file", "path": str(path), "exists": path.exists(), "configured": configured, "managed": managed}


def _install_opencode(ctx: McpClientContext) -> dict[str, Any]:
    path = _opencode_config_path(ctx)
    data = _read_json(path)
    mcp = data.setdefault("mcp", {})
    existing = mcp.get("lovart")
    if existing and not (isinstance(existing, dict) and existing.get("managed_by") == "lovart-reverse") and not ctx.force:
        raise ConfigConflictError("existing unmanaged OpenCode Lovart MCP config", {"client": "opencode", "path": str(path), "recommended_actions": ["rerun with --force", "edit the config manually"]})
    mcp["lovart"] = {"type": "local", "command": [str(ctx.lovart_path), "mcp"], "enabled": True, "managed_by": "lovart-reverse"}
    text = json.dumps(data, ensure_ascii=False, indent=2) + "\n"
    return _write_config_result("opencode", path, text, ctx, preview={"json": mcp["lovart"]})


def _install_cli_client(client: str, command: list[str], ctx: McpClientContext) -> dict[str, Any]:
    executable = command[0]
    available = shutil.which(executable) is not None
    result = {"client": client, "type": "command", "available": available, "command": command, "manual_command": _shell_join(command), "changed": False}
    if ctx.dry_run or not available:
        result["status"] = "dry_run" if ctx.dry_run else "manual_required"
        return result
    completed = subprocess.run(command, capture_output=True, text=True, check=False)
    result.update({"status": "configured" if completed.returncode == 0 else "failed", "changed": completed.returncode == 0, "returncode": completed.returncode, "stderr": completed.stderr[-2000:], "stdout": completed.stdout[-2000:]})
    if completed.returncode != 0:
        raise LovartError("mcp_config_failed", f"{client} MCP configuration failed", result, 2)
    return result


def _command_status(client: str, command: list[str]) -> dict[str, Any]:
    available = shutil.which(command[0]) is not None
    return {"client": client, "type": "command", "available": available, "configured": None, "manual_command": _shell_join(command)}


def _openclaw_payload(ctx: McpClientContext) -> str:
    return json.dumps({"command": str(ctx.lovart_path), "args": ["mcp"]}, separators=(",", ":"))


def _write_config_result(client: str, path: Path, text: str, ctx: McpClientContext, *, preview: dict[str, Any]) -> dict[str, Any]:
    backup = _backup_path(path)
    result = {"client": client, "type": "file", "path": str(path), "backup": str(backup), "changed": False, "dry_run": ctx.dry_run, "preview": preview}
    if ctx.dry_run:
        result["status"] = "dry_run"
        return result
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists():
        backup.write_text(path.read_text())
    path.write_text(text)
    result.update({"changed": True, "status": "configured", "backup_created": path.exists() and backup.exists()})
    return result


def _backup_path(path: Path) -> Path:
    stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    return path.with_name(f"{path.name}.{stamp}.bak")


def _read_text(path: Path) -> str:
    try:
        return path.read_text()
    except FileNotFoundError:
        return ""


def _read_json(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text())
    except json.JSONDecodeError as exc:
        raise LovartError("config_invalid", "config file is not valid JSON", {"path": str(path), "error": str(exc)}, 2)
    if not isinstance(data, dict):
        raise LovartError("config_invalid", "config file root must be an object", {"path": str(path)}, 2)
    return data


def _toml_string(value: str) -> str:
    return value.replace("\\", "\\\\").replace('"', '\\"')


def _shell_join(command: list[str]) -> str:
    return " ".join(_shell_quote(part) for part in command)


def _shell_quote(value: str) -> str:
    if not value or any(ch.isspace() or ch in "'\"\\$`{}[]" for ch in value):
        return "'" + value.replace("'", "'\"'\"'") + "'"
    return value

File: mcp/configure.go
package mcp

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/aporicho/lovart/internal/envelope"
	"github.com/aporicho/lovart/internal/errors"
)

const managedMarker = "# Managed by lovart"

var supportedMCPClients = []string{"codex", "claude", "opencode", "openclaw"}

var (
	lookPath    = exec.LookPath
	runCommand  = runCommandDefault
	nowUTC      = func() time.Time { return time.Now().UTC() }
	userHomeDir = os.UserHomeDir
)

// ConfigOptions controls MCP client status and install operations.
type ConfigOptions struct {
	Clients    string
	LovartPath string
	Home       string
	DryRun     bool
	Yes        bool
	Force      bool
}

type configContext struct {
	lovartPath string
	home       string
	dryRun     bool
	yes        bool
	force      bool
}

type commandResult struct {
	ReturnCode int
	Stdout     string
	Stderr     string
}

// Install configures selected local MCP clients to use this lovart binary.
func Install(opts ConfigOptions) envelope.Envelope {
	if !opts.DryRun && !opts.Yes {
		return envelope.Err(errors.CodeInputError, "--yes is required for mcp install", map[string]any{
			"recommended_actions": []string{"rerun with --yes", "rerun with --dry-run to preview changes"},
		})
	}
	ctx, err := newConfigContext(opts)
	if err != nil {
		return envelope.Err(errors.CodeInputError, "mcp install setup failed", map[string]any{"error": err.Error()})
	}
	selected, err := selectMCPClients(clientSpec(opts.Clients), ctx, false)
	if err != nil {
		return envelope.Err(errors.CodeInputError, "unknown MCP client", map[string]any{
			"error":             err.Error(),
			"supported_clients": supportedMCPClients,
		})
	}
	results := make([]map[string]any, 0, len(selected))
	for _, client := range selected {
		result, err := installClient(client, ctx)
		if err != nil {
			return configErrorEnvelope(err)
		}
		results = append(results, result)
	}
	return okLocal(map[string]any{
		"lovart_path":            ctx.lovartPath,
		"dry_run":                ctx.dryRun,
		"force":                  ctx.force,
		"mcp_clients_requested":  clientSpec(opts.Clients),
		"mcp_clients_selected":   selected,
		"supported_mcp_clients":  supportedMCPClients,
		"results":                results,
		"recommended_next_steps": []string{"run `lovart mcp status`", "restart your MCP client"},
	}, true)
}

// Uninstall removes Lovart MCP config from selected local clients.
func Uninstall(opts ConfigOptions) envelope.Envelope {
	if !opts.DryRun && !opts.Yes {
		return envelope.Err(errors.CodeInputError, "--yes is required for mcp uninstall", map[string]any{
			"recommended_actions": []string{"rerun with --yes", "rerun with --dry-run to preview changes"},
		})
	}
	ctx, err := newConfigContext(opts)
	if err != nil {
		return envelope.Err(errors.CodeInputError, "mcp uninstall setup failed", map[string]any{"error": err.Error()})
	}
	selected, err := selectMCPClientsForUninstall(clientSpec(opts.Clients), ctx)
	if err != nil {
		return envelope.Err(errors.CodeInputError, "unknown MCP client", map[string]any{
			"error":             err.Error(),
			"supported_clients": supportedMCPClients,
		})
	}
	results := make([]map[string]any, 0, len(selected))
	for _, client := range selected {
		result, err := uninstallClient(client, ctx)
		if err != nil {
			return configErrorEnvelope(err)
		}
		results = append(results, result)
	}
	return okLocal(map[string]any{
		"lovart_path":            ctx.lovartPath,
		"dry_run":                ctx.dryRun,
		"force":                  ctx.force,
		"mcp_clients_requested":  clientSpec(opts.Clients),
		"mcp_clients_selected":   selected,
		"supported_mcp_clients":  supportedMCPClients,
		"results":                results,
		"recommended_next_steps": []string{"restart your MCP client"},
	}, true)
}

// ConfigStatus reports selected local MCP client configuration state.
func ConfigStatus(opts ConfigOptions) (map[string]any, error) {
	ctx, err := newConfigContext(opts)
	if err != nil {
		return nil, err
	}
	selected, err := selectMCPClients(clientSpec(opts.Clients), ctx, true)
	if err != nil {
		return nil, err
	}
	statuses := make([]map[string]any, 0, len(selected))
	for _, client := range selected {
		statuses = append(statuses, statusForClient(client, ctx))
	}
	return map[string]any{
		"lovart_path":           ctx.lovartPath,
		"mcp_clients_requested": clientSpec(opts.Clients),
		"mcp_clients_selected":  selected,
		"supported_clients":     supportedMCPClients,
		"clients":               statuses,
	}, nil
}

func newConfigContext(opts ConfigOptions) (configContext, error) {
	path, err := resolveLovartPath(opts.LovartPath)
	if err != nil {
		return configContext{}, err
	}
	home := opts.Home
	if home == "" {
		home, err = userHomeDir()
		if err != nil {
			return configContext{}, err
		}
	}
	home, err = filepath.Abs(expandHome(home))
	if err != nil {
		return configContext{}, err
	}
	return configContext{lovartPath: path, home: home, dryRun: opts.DryRun, yes: opts.Yes, force: opts.Force}, nil
}

func resolveLovartPath(path string) (string, error) {
	if path == "" {
		path = lovartPath()
	}
	abs, err := filepath.Abs(expandHome(path))
	if err != nil {
		return "", err
	}
	return abs, nil
}

func expandHome(path string) string {
	if path == "~" {
		if home, err := userHomeDir(); err == nil {
			return home
		}
		return path
	}
	if strings.HasPrefix(path, "~/") {
		if home, err := userHomeDir(); err == nil {
			return filepath.Join(home, strings.TrimPrefix(path, "~/"))
		}
	}
	return path
}

func clientSpec(value string) string {
	if strings.TrimSpace(value) == "" {
		return "auto"
	}
	return strings.TrimSpace(strings.ToLower(value))
}

func selectMCPClients(spec string, ctx configContext, includeMissing bool) ([]string, error) {
	switch spec {
	case "none":
		return []string{}, nil
	case "all":
		return append([]string(nil), supportedMCPClients...), nil
	case "auto":
		var selected []string
		for _, client := range supportedMCPClients {
			if clientDetected(client, ctx) {
				selected = append(selected, client)
			}
		}
		if len(selected) > 0 {
			return selected, nil
		}
		if includeMissing {
			return append([]string(nil), supportedMCPClients...), nil
		}
		return []string{"codex"}, nil
	default:
		parts := strings.Split(spec, ",")
		seen := map[string]bool{}
		var selected []string
		var unknown []string
		for _, part := range parts {
			client := strings.TrimSpace(strings.ToLower(part))
			if client == "" {
				continue
			}
			if !isSupportedClient(client) {
				unknown = append(unknown, client)
				continue
			}
			if !seen[client] {
				seen[client] = true
				selected = append(selected, client)
			}
		}
		if len(unknown) > 0 {
			sort.Strings(unknown)
			return nil, fmt.Errorf("unknown clients: %s", strings.Join(unknown, ", "))
		}
		return selected, nil
	}
}

func selectMCPClientsForUninstall(spec string, ctx configContext) ([]string, error) {
	if spec != "auto" {
		return selectMCPClients(spec, ctx, false)
	}
	var selected []string
	for _, client := range supportedMCPClients {
		if clientDetected(client, ctx) {
			selected = append(selected, client)
		}
	}
	return selected, nil
}

func isSupportedClient(client string) bool {
	for _, supported := range supportedMCPClients {
		if client == supported {
			return true
		}
	}
	return false
}

func clientDetected(client string, ctx configContext) bool {
	switch client {
	case "codex":
		return pathExists(filepath.Join(ctx.home, ".codex"))
	case "claude":
		return commandAvailable("claude")
	case "opencode":
		return commandAvailable("opencode") || pathExists(filepath.Join(ctx.home, ".config", "opencode"))
	case "openclaw":
		return commandAvailable("openclaw")
	default:
		return false
	}
}

func statusForClient(client string, ctx configContext) map[string]any {
	switch client {
	case "codex":
		return codexStatus(ctx)
	case "claude":
		return commandStatus("claude", claudeCommand(ctx))
	case "opencode":
		return opencodeStatus(ctx)
	case "openclaw":
		return commandStatus("openclaw", openclawCommand(ctx))
	default:
		return map[string]any{"client": client, "status": "unknown"}
	}
}

func installClient(client string, ctx configContext) (map[string]any, error) {
	switch client {
	case "codex":
		return installCodex(ctx)
	case "claude":
		return installClaude(ctx)
	case "opencode":
		return installOpenCode(ctx)
	case "openclaw":
		return installCommandClient("openclaw", openclawCommand(ctx), ctx)
	default:
		return nil, configInputError{Message: "unknown MCP client", Details: map[string]any{"client": client}}
	}
}

func uninstallClient(client string, ctx configContext) (map[string]any, error) {
	switch client {
	case "codex":
		return uninstallCodex(ctx)
	case "claude":
		return uninstallCommandClient("claude", claudeRemoveCommand(), ctx)
	case "opencode":
		return uninstallOpenCode(ctx)
	case "openclaw":
		return map[string]any{
			"client":             "openclaw",
			"type":               "command",
			"available":          commandAvailable("openclaw"),
			"status":             "manual_required",
			"changed":            false,
			"recommended_action": "remove the Lovart MCP server from OpenClaw manually",
		}, nil
	default:
		return nil, configInputError{Message: "unknown MCP client", Details: map[string]any{"client": client}}
	}
}

func commandStatus(client string, command []string) map[string]any {
	return map[string]any{
		"client":         client,
		"type":           "command",
		"available":      commandAvailable(command[0]),
		"configured":     nil,
		"manual_command": shellJoin(command),
		"command":        command,
	}
}

func installCommandClient(client string, command []string, ctx configContext) (map[string]any, error) {
	result := commandStatus(client, command)
	result["changed"] = false
	if ctx.dryRun {
		result["status"] = "dry_run"
		return result, nil
	}
	if !commandAvailable(command[0]) {
		result["status"] = "manual_required"
		return result, nil
	}
	completed, err := runCommand(command)
	if err != nil {
		result["status"] = "failed"
		result["error"] = err.Error()
		return nil, configCommandError{Message: client + " MCP configuration failed", Details: result}
	}
	result["returncode"] = completed.ReturnCode
	result["stdout"] = tail(completed.Stdout, 2000)
	result["stderr"] = tail(completed.Stderr, 2000)
	if completed.ReturnCode != 0 {
		result["status"] = "failed"
		return nil, configCommandError{Message: client + " MCP configuration failed", Details: result}
	}
	result["status"] = "configured"
	result["changed"] = true
	return result, nil
}

func installClaude(ctx configContext) (map[string]any, error) {
	command := claudeCommand(ctx)
	result := commandStatus("claude", command)
	result["changed"] = false
	if ctx.dryRun {
		result["status"] = "dry_run"
		return result, nil
	}
	if !commandAvailable(command[0]) {
		result["status"] = "manual_required"
		return result, nil
	}

	completed, err := runCommand(command)
	recordCommandResult(result, completed)
	if err != nil {
		result["status"] = "failed"
		result["error"] = err.Error()
		return nil, configCommandError{Message: "claude MCP configuration failed", Details: result}
	}
	if completed.ReturnCode == 0 {
		result["status"] = "configured"
		result["changed"] = true
		return result, nil
	}
	if !commandAlreadyExists(completed) {
		result["status"] = "failed"
		return nil, configCommandError{Message: "claude MCP configuration failed", Details: result}
	}

	result["configured"] = true
	if !ctx.force {
		result["status"] = "already_exists"
		result["recommended_action"] = "rerun with --force to replace the existing Claude Lovart MCP server"
		return result, nil
	}

	result["initial_returncode"] = completed.ReturnCode
	result["initial_stdout"] = tail(completed.Stdout, 2000)
	result["initial_stderr"] = tail(completed.Stderr, 2000)
	removeCommand := claudeRemoveCommand()
	result["remove_command"] = removeCommand
	removed, err := runCommand(removeCommand)
	result["remove_returncode"] = removed.ReturnCode
	result["remove_stdout"] = tail(removed.Stdout, 2000)
	result["remove_stderr"] = tail(removed.Stderr, 2000)
	if err != nil {
		result["status"] = "failed"
		result["error"] = err.Error()
		return nil, configCommandError{Message: "claude MCP replacement failed", Details: result}
	}
	if removed.ReturnCode != 0 {
		result["status"] = "failed"
		return nil, configCommandError{Message: "claude MCP replacement failed", Details: result}
	}

	retried, err := runCommand(command)
	recordCommandResult(result, retried)
	if err != nil {
		result["status"] = "failed"
		result["error"] = err.Error()
		return nil, configCommandError{Message: "claude MCP configuration failed", Details: result}
	}
	if retried.ReturnCode != 0 {
		result["status"] = "failed"
		return nil, configCommandError{Message: "claude MCP configuration failed", Details: result}
	}
	result["status"] = "configured"
	result["changed"] = true
	result["replaced"] = true
	return result, nil
}

func uninstallCommandClient(client string, command []string, ctx configContext) (map[string]any, error) {
	result := commandStatus(client, command)
	result["changed"] = false
	if ctx.dryRun {
		result["status"] = "dry_run"
		return result, nil
	}
	if !commandAvailable(command[0]) {
		result["status"] = "manual_required"
		return result, nil
	}
	completed, err := runCommand(command)
	if err != nil {
		result["status"] = "failed"
		result["error"] = err.Error()
		return nil, configCommandError{Message: client + " MCP removal failed", Details: result}
	}
	result["returncode"] = completed.ReturnCode
	result["stdout"] = tail(completed.Stdout, 2000)
	result["stderr"] = tail(completed.Stderr, 2000)
	if completed.ReturnCode != 0 {
		result["status"] = "failed"
		return nil, configCommandError{Message: client + " MCP removal failed", Details: result}
	}
	result["status"] = "removed"
	result["changed"] = true
	return result, nil
}

func recordCommandResult(result map[string]any, completed commandResult) {
	result["returncode"] = completed.ReturnCode
	result["stdout"] = tail(completed.Stdout, 2000)
	result["stderr"] = tail(completed.Stderr, 2000)
}

func commandAlreadyExists(completed commandResult) bool {
	text := strings.ToLower(completed.Stdout + "\n" + completed.Stderr)
	return strings.Contains(text, "already exists")
}

func claudeCommand(ctx configContext) []string {
	return []string{"claude", "mcp", "add", "--transport", "stdio", "lovart", "--scope", "user", "--", ctx.lovartPath, "mcp"}
}

func claudeRemoveCommand() []string {
	return []string{"claude", "mcp", "remove", "--scope", "user", "lovart"}
}

func openclawCommand(ctx configContext) []string {
	payload, _ := json.Marshal(map[string]any{"command": ctx.lovartPath, "args": []string{"mcp"}})
	return []string{"openclaw", "mcp", "set", "lovart", string(payload)}
}

func commandAvailable(name string) bool {
	_, err := lookPath(name)
	return err == nil
}

func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func runCommandDefault(command []string) (commandResult, error) {
	cmd := exec.Command(command[0], command[1:]...)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	err := cmd.Run()
	result := commandResult{Stdout: stdout.String(), Stderr: stderr.String()}
	if exitErr, ok := err.(*exec.ExitError); ok {
		result.ReturnCode = exitErr.ExitCode()
		return result, nil
	}
	if err != nil {
		return result, err
	}
	result.ReturnCode = 0
	return result, nil
}

func shellJoin(args []string) string {
	quoted := make([]string, 0, len(args))
	for _, arg := range args {
		if arg == "" || strings.ContainsAny(arg, " \t\n'\"\\$`") {
			quoted = append(quoted, "'"+strings.ReplaceAll(arg, "'", "'\\''")+"'")
		} else {
			quoted = append(quoted, arg)
		}
	}
	return strings.Join(quoted, " ")
}

func tail(value string, limit int) string {
	if len(value) <= limit {
		return value
	}
	return value[len(value)-limit:]
}
